### PR TITLE
Adding req_id generation

### PR DESCRIPTION
**Pull request by JeffAshton**
As recommended by the guys a Joyent ( http://trentm.com/talk-bunyan-in-prod/#/21 ), tagging requests with a unique id makes it dead simple to query for all the logs pertaining to an individual request.

### CODE REVIEW

**Comment by leedm777** [comment]
I always thought of this as something that should be outside of `sails-hook-bunyan`. We do this, but we have a different middleware that adds the id to the request, and a custom serializer that includes `req.id` when serializing the request.

**Comment by leedm777** [comment]
@JeffAshtonCT OTOH, after some thought, logging `req` with every message seems overkill. Might be preferable to log the `req` just once, and have every other message just reference the request id. Thoughts?

**Comment by JeffAshton** [comment]
Logging the request at the beginning and the response at the end makes sense.  But only if the request id is present in all messages within.  

At the same time, having the request info in every message is really nice too for searching. If I wanted to search for a particular error in /example /buggy/route, then the request info is great to have.  Might make sense to just have a config for this.  But if no one is asking, hold off.

I like what this guy is proposing too https://brandur.org/request-ids#multiple), of tracking multiple request ids, effectively producing a service call stack.  

I feel request id makes sense in this project as it's designed for the benefit of logging.  I can't think of a good reason to not do it, which is why i opted for it defaulting to true.  Will likely add support for X-Request-Id over the weekend and send you another pull request.

File: README.md [hunk 1]
```diff
@@ -23,6 +23,9 @@ module.exports.bunyan = {
   /** If true, a child logger is injected on each request */
   injectRequestLogger: true,
 
+  /** If true and injectRequestLogger is enabled, then generate and attach a unique req_id to each request logger */
```
Comment (leedm777) [inline README.md]: margins.

File: index.js [hunk 2]
```diff
@@ -155,7 +162,17 @@ module.exports = function(sails) {
       before: {
         '/*': function(req, res, next) {
           if (injectRequestLogger) {
-            req.log = _this.logger.child({req: req}, true);
+
+            var options = {req: req};
```
Comment (leedm777) [inline index.js]: Please add a `addReqToChildLogger` config options, to make this optional.

File: index.js [hunk 3]
```diff
@@ -155,7 +162,17 @@ module.exports = function(sails) {
...
+
+            if (generateRequestId) {
+
+              // jscs: disable
+              options.req_id = uuid.v4(); // jshint ignore:line
```
Comment (leedm777) [inline index.js]: What do you think about including this in the `req` itself? `req.id = options.req_id = uuid.v4()`. That way the request id is accessible from other contexts, like being able to pass it along with other requests throughout the system.

File: index.js [hunk 4]
```diff
@@ -2,6 +2,7 @@
 'use strict';
 
 var bunyan = require('bunyan');
+var uuid = require('uuid');
```
Comment (leedm777) [inline index.js:5]: `npm i --save uuid`

**Comment by leedm777** [comment]
1. Fix/add items noted above
2. Rebase against master and squash into a single commit

Thanks for the improvement!

File: index.js (hunk 3 above)
Comment (JeffAshton) [inline index.js]: Yeah that's a good idea.  Will add.  Necessary to pass along to other service calls.

File: README.md (hunk 1 above)
Comment (JeffAshton) [inline README.md]: What line length are you targetting?  I had it all on one line before the static analysis tool complained.  ;)

Comment (leedm777) [inline README.md]: 80 columns

**Comment by JeffAshton** [comment]
I started looking at options for supporting a couple different scenarios,

1) attach a unique id to every request
2) support heroku style (https://devcenter.heroku.com/articles/http-request-id), where a request id can be specified via X-REQUEST-ID header, otherwise generated
3) support request id stack via X-REQUEST-ID header, as proposed by https://brandur.org/request-ids#multiple

I think it should be this project's goal to setup the Logger on every request (which you have), and make it easy to assign that request id to the Logger options.  Played with a couple different configuration scenarios, and here's my proposal

``` js
module.exports.log = {

    /** If true, a child logger is injected on each request */
    injectRequestLogger: true,

    /**
     * Gets or generates a unique id for the request, and attaches it
     * to the request logger's options. If no id is returned, then the
     * request logger is unmodified.
     *
     * The default provider returns a new UUID v4 and attaches
     * it to req.id
     * 
     * Note: Only called if injectRequestLogger is true.
     */
    requestIdProvider: function( req ) {

        if( ! req.id ) {
            req.id = uuid.v4();
        }
        return req.id;
    },

    /** Name of the request id property assigned to the request logger */
    requestIdProperty: 'req_id'
}
```

This design should support all 3 use cases above, and give developer's joyent's recommended practice by default.  Thoughts?

**Comment by JeffAshton** [comment]
Opened new pull request with squashed changes.  I like how the requestIdProvider turned out.  Makes it easy to do the recommended minimal id per request, but still gives the flexibility to extend outside this project.

**Pull request closed without merging**